base/array.jl

```diff
@@ -237,16 +237,29 @@ function _collect(cont, itr, ::HasEltype, isz::SizeUnknown)
     return a
 end
 
+if isdefined(:Base)
+@eval function _default_eltype(itrt::ANY)
+    $(Expr(:meta, :pure))
+    try
+        rt = return_types(first, (itrt,))
```

Comment:
I would be happy for it to have a better interface. I think there's even some some valid uses of a static_typeof-like construct for optimizing closures (allowing the Boxes to be more accurately typed, but invisible to the user). I just don't think that `return_types` is capable of it, since you really need to return a type-inference edge, not a value, and we don't have first-class edges in type-inference. I'm not sure if you can have a first-class edge (it definitely permits breaking the monotonicity constraints on termination). It might be roughly correct for the edge to be represented through a closure?

```julia
return_types(f, at) do rt
    update!(?, rt)
end::Void
```

Alternatively, maybe you could return some sort of "promise" object that tracks the edge. I'm not really sure how that would look however.